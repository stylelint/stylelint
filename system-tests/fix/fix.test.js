"use strict";

const _ = require("lodash");
const cpFile = require("cp-file");
const del = require("del");
const fs = require("fs");
const os = require("os");
const path = require("path");
const pify = require("pify");
const stylelint = require("../../lib");
const systemTestUtils = require("../systemTestUtils");

describe("fix", () => {
  let tmpDir;
  let stylesheetPath;

  beforeEach(() => {
    tmpDir = os.tmpdir();
    stylesheetPath = path.join(tmpDir, `stylesheet-${_.uniqueId()}.css`);

    return cpFile(path.join(__dirname, "stylesheet.css"), stylesheetPath);
  });

  afterEach(() => {
    return del(stylesheetPath, { force: true });
  });

  it("expected stylelint results", () => {
    return stylelint
      .lint({
        files: [stylesheetPath],
        configFile: systemTestUtils.caseConfig("fix"),
        fix: true
      })
      .then(output => {
        // Remove the path to tmpDir
        const cleanedResults = output.results.map(r =>
          Object.assign({}, r, { source: "stylesheet.css" })
        );

        expect(systemTestUtils.prepResults(cleanedResults)).toMatchSnapshot();
      });
  });

  it("expected fixes to PostCSS Result", () => {
    return stylelint
      .lint({
        files: [stylesheetPath],
        configFile: systemTestUtils.caseConfig("fix"),
        fix: true
      })
      .then(output => {
        const result = output.results[0]._postcssResult;

        expect(result.root.toString(result.opts.syntax)).toMatchSnapshot();
      });
  });

  it("overwrites the original file", () => {
    return stylelint
      .lint({
        files: [stylesheetPath],
        configFile: systemTestUtils.caseConfig("fix"),
        fix: true
      })
      .then(output => {
        const result = output.results[0]._postcssResult;

        return pify(fs.readFile)(stylesheetPath, "utf8").then(fileContent => {
          expect(fileContent).toBe(result.root.toString(result.opts.syntax));
        });
      });
  });

  it("doesn't write to ignored file", () => {
    return stylelint
      .lint({
        files: [stylesheetPath],
        config: {
          ignoreFiles: stylesheetPath,
          rules: {
            "at-rule-name-case": "lower",
            "comment-empty-line-before": "always",
            "comment-no-empty": true
          }
        },
        fix: true
      })
      .then(() => {
        return pify(fs.readFile)(stylesheetPath, "utf8").then(newFile => {
          return pify(fs.readFile)(
            path.join(__dirname, "stylesheet.css"),
            "utf8"
          ).then(oldFile => {
            expect(newFile).toBe(oldFile);
          });
        });
      });
  });

  it("outputs fixed code when input is code string", () => {
    return stylelint
      .lint({
        code: "  a { color: red; }",
        config: {
          rules: {
            indentation: 2
          }
        },
        fix: true
      })
      .then(result => {
        expect(result.output).toBe("a { color: red; }");
      });
  });
});
