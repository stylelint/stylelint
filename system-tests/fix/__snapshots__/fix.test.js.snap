// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fix expected fixes to PostCSS Result 1`] = `
"a { @extend placeholder; }
a {} @media {}
@charset 'UTF-8';
a { color: #000; }
a {}

/* comment */

/**/

/*comment*/

/* TODO: */
@custom-media --bar (min-width: 30em);
a {
  & .foo { /* 1 */
    &__foo { /* 2 */
      & > .bar {} /* 3 */
    }
  }
}
"
`;

exports[`fix expected stylelint results 1`] = `
Array [
  Object {
    "deprecations": Array [],
    "errored": true,
    "ignored": undefined,
    "invalidOptionWarnings": Array [],
    "parseErrors": Array [],
    "source": "../stylesheet.css",
    "warnings": Array [
      Object {
        "column": 1,
        "line": 7,
        "rule": "comment-no-empty",
        "severity": "error",
        "text": "Unexpected empty comment (comment-no-empty)",
      },
    ],
  },
]
`;
